Add tutorial for using Pyrokinetics on HPC systems

LiamPattinson opened the pull request:
A while back, I recall talking about a potential feature that could automatically dispatch gyrokinetics jobs to a remote HPC system (although I don't think an issue was raised for this). Having looked into it a bit, I think it would be very difficult to design a generic tool for doing this, so instead I've written a short guide on how it could be achieved for a simple toy problem. It first explains how to install Pyrokinetics on an HPC machine, and then gives a very rough overview of two tools that can be used to automate Pyrokinetics runs:

- [QCG-PilotJob](https://qcg-pilotjob.readthedocs.io/en/develop/): A job scheduler that runs within a single Slurm job. Allows you to queue up a large number of jobs and manages resource allocation for you.
- [HPC Rocket](https://svenmarcus.github.io/hpc-rocket/): A tool for dispatching jobs to remote HPC machines without the need to SSH in. Mostly intended as a CI tool, but it can used as a regular command line tool too.

Hopefully this might be useful to somebody! If not, I've at least had some fun playing around with new tools.

LiamPattinson committed:
Add tutorial for using Pyrokinetics on HPC systems

d7919 commented on docs/tutorials/using_pyrokinetics_on_hpc_systems.rst:
```diff
@@ -0,0 +1,395 @@
+
+.. code-block:: bash
+
+   $ scp pyrokinetics_template.in *.geqdsk username@hpc.com:scratch/pyrokinetics/
```
```suggestion
   $ scp pyrokinetics_template.in *.geqdsk username@hpc.machine.com:scratch/pyrokinetics/
```

d7919 commented on docs/tutorials/using_pyrokinetics_on_hpc_systems.rst:
```diff
@@ -0,0 +1,395 @@
+   # Perform runs
+   # ------------
+
+   srun $gs2 tdotp_lowq0.geqdsk.d/tdotp_lowq0.in
```
Note slurm has a (sysadmin configurable) maximum number of srun commands which can appear in a single job submission. The default is something like 40,000 so unlikely to hit this, but might be worth a mention somewhere?

d7919 commented on docs/tutorials/using_pyrokinetics_on_hpc_systems.rst:
```diff
@@ -0,0 +1,395 @@
+
+.. code-block:: bash
+
+   $ scp username@hpc.com:scratch/pyrokinetics/tdotp_lowq0.geqdsk.d/tdotp_lowq0.out.nc .
```
```suggestion
   $ scp username@hpc.machine.com:scratch/pyrokinetics/tdotp_lowq0.geqdsk.d/tdotp_lowq0.out.nc .
```

d7919 commented on docs/tutorials/using_pyrokinetics_on_hpc_systems.rst:
```diff
@@ -0,0 +1,395 @@
+   #SBATCH --mail-user=me@email.com # Where to send mail reporting on the job
+   #SBATCH --mail-type=END,FAIL     # Mail events (NONE, BEGIN, END, FAIL, ALL)
+   #SBATCH --nodes=1                # Number of compute nodes to run on
+   #SBATCH --ntasks-per-node=24     # Number of MPI processes to spawn per node
```
Difference wrt simple case -- might be worth keeping this the same as before to avoid "noise"

d7919 replied:
Ah I see from the python that you're asking each job to use 8 cores so I guess  the intention here is to run all three cases at the same time -- might be worth some text up front to indicate (e.g. before the python script) to say what approach to running you're targeting here?

d7919 commented on docs/tutorials/using_pyrokinetics_on_hpc_systems.rst:
```diff
@@ -0,0 +1,395 @@
+
+.. code-block:: yaml
+
+   host: hpc.com
```
```suggestion
   host: hpc.machine.com
```

d7919 commented on docs/tutorials/using_pyrokinetics_on_hpc_systems.rst:
```diff
@@ -0,0 +1,395 @@
+
+   host: hpc.com
+   user: username
+   password: $PASSWORD
```
This feels a little dubious (and presumably won't work with systems requiring MFA?) -- might it be worth documenting how to set up a ControlMaster config (e.g. https://support.york.ac.uk/s/article/How-to-create-a-ControlMaster-SSH-session-so-that-you-only-need-to-authenticate-once-for-multiple-SSH-sessions) so that users just authenticate once and then hpc-rocket can reuse this?

LiamPattinson committed:
HPC docs: use consistent machine name

LiamPattinson committed:
HPC docs: consistent resource usage

LiamPattinson committed:
HPC docs: Add srun details

LiamPattinson committed:
HPC docs: Add SSH keys info

LiamPattinson committed:
HPC docs: minor edit to srun details

LiamPattinson committed:
HPC docs: Use less memory in examples

LiamPattinson commented:
Thanks for the review! I've fixed up all of the inconsistencies, added some details on `srun` usage, and added some info on using SSH keys instead of supplying a password to HPC Rocket (I think a password might make more sense when you're running it as a CI job, as you can just save your password to a GitHub secret, but you're right that SSH keys are preferable in all other cases).

d7919 commented on docs/tutorials/using_pyrokinetics_on_hpc_systems.rst:
```diff
@@ -385,8 +399,34 @@ We can set up a remote job run by creating a new directory, adding the scripts i
+Including ``-fN`` will result in your session sitting idle in the background, while
+``-o ServerAliveInternal=30`` will ping the server every 30 seconds to ensure your
+connection stays alive. If you aren't using multi-factor authentication, it should be
+sufficient to just provide a password or private key file, and there's no need to create
```
Presumably one would need to unlock the private key file (e.g. with ssh-add) if it is not passwordless.

LiamPattinson committed:
HPC docs: Add SSH keys password solution

LiamPattinson commented:
I've added a few lines about unlocking private key passwords beforehand. I'd actually been having issues with my keychain for a while, and this seems to have fixed it, so cheers for the prompt!

bpatel2107 commented:
This looks fantastic! @whornsby82 @lorenzozanisi FYI may be of interest

Is it worth including info on how to pull the output data from the HPC cluster? I think that would close the whole loop though it is a bit out of the scope.

LiamPattinson committed:
HPC job: Add more HPC Rocket details

LiamPattinson commented:
There's a bit in the 'Simple usage' section about `scp`-ing output files back, but I guess I never explained how it works for HPC Rocket. I've added a small section to end that explains how to ensure HPC Rocket pulls files back to you after the batch job has finished.
